File: frontend/types/dailyCost.ts
/**
 * Type contracts for the Daily Costs, Cost Analytics and Cost Reports pages.
 *
 * Every money value arrives from the API as a decimal string (`"1234.50"`) and
 * is sent back the same way, so nothing is ever rounded in the browser: the
 * server prices a day and the page only displays what came back.
 */

import type { ChargingBasis, QuantityUnit } from '~/types/afe'

export type { ChargingBasis, QuantityUnit }

export type DailyStatus = 'draft' | 'submitted'
export type ReconciliationStatus = 'pending' | 'reconciled'

/** The four consumable categories entered on a day. */
export type ConsumableCategory = 'mud_chemical' | 'fuel' | 'cement_additive' | 'drill_bit'

export const CONSUMABLE_CATEGORIES: ConsumableCategory[] = [
  'mud_chemical',
  'fuel',
  'cement_additive',
  'drill_bit',
]

export const CONSUMABLE_LABELS: Record<ConsumableCategory, string> = {
  mud_chemical: 'Mud Chemicals',
  fuel: 'Fuel',
  cement_additive: 'Cement Additives',
  drill_bit: 'Drill Bits',
}

/** How each consumable category is entered on a day. */
export const CONSUMABLE_ENTRY_HINTS: Record<ConsumableCategory, string> = {
  mud_chemical: 'Pick the chemical from Master Data and enter the quantity used — the unit rate is captured from the catalogue.',
  fuel: 'Enter the usage only — the fuel unit rate is captured from the AFE cost estimate.',
  cement_additive: 'Enter the total consumption cost for the selected section, phase and sub activity.',
  drill_bit: 'Pick the drill bit and enter the number used — the unit rate is captured from the catalogue.',
}

/** Categories where the AFE amount is a one-time charge, never multiplied. */
export const ONE_TIME_LABEL = 'One-time charge — not multiplied by hours or days'

/** Quantity limits enforced by the engine. */
export const MAX_HOURS = 24
export const MAX_DAYS = 1

// ---------------------------------------------------------------------------
// Context (what the well + AFE provide)
// ---------------------------------------------------------------------------

export interface RateCardRate {
  category: string
  unit_rate: string
  [key: string]: unknown
}

export interface RateCardSectionRate {
  section_id: number
  phase_id: number | null
  amount: string
  section_label?: string | null
  [key: string]: unknown
}

/** One service of the selected AFE, with everything the daily page prices from. */
export interface RateCardService {
  service_id: number
  afe_line_id: number | null
  service_code: string
  service_name: string
  provider_type: string
  charging_basis: ChargingBasis
  per_service_amount: string
  section_id: number | null
  phase_id: number | null
  rates: RateCardRate[]
  section_rates: RateCardSectionRate[]
}

export interface DailyCostAfeOption {
  id: number
  afe_code: string
  afe_name: string
  afe_type?: string
  status?: string
  estimated_total?: string
  [key: string]: unknown
}

export interface DailyCostSubActivity {
  id: number
  sub_activity_code: string
  sub_activity_name: string
  activity_id: number | null
  activity_code?: string | null
  activity_name?: string | null
  [key: string]: unknown
}

export interface WellConfigurationSectionPhase {
  phase_id: number
  phase_code: string
  phase_name: string
  days: string
  [key: string]: unknown
}

export interface WellConfigurationSection {
  section_id: number
  section_code: string
  section_name: string
  from_depth: string | null
  to_depth: string | null
  phases: WellConfigurationSectionPhase[]
  [key: string]: unknown
}

export interface WellConfiguration {
  sections: WellConfigurationSection[]
  total_depth: string
  depth_unit: string
  total_days: string
  [key: string]: unknown
}

export interface DailyCostContext {
  well_id: number
  well_code: string
  well_name: string
  rig_id: number | null
  rig_code: string | null
  rig_name: string | null
  depth_unit: string
  well_configuration: WellConfiguration | null
  afes: DailyCostAfeOption[]
  sub_activities: DailyCostSubActivity[]
  rate_card: RateCardService[]
  afe_id: number | null
  /** Fuel unit rate captured on the selected AFE cost estimate. */
  fuel_rate: string
  /** The selected AFE's estimated total, to show what the day leaves behind. */
  afe_estimated_total: string
  warnings: string[]
}

// ---------------------------------------------------------------------------
// Day lines: the payloads and the read models
// ---------------------------------------------------------------------------

export interface DailyServiceLineIn {
  service_id: number | null
  charging_basis?: ChargingBasis | null
  charge_category?: string | null
  afe_line_id?: number | null
  section_id?: number | null
  phase_id?: number | null
  sub_activity_id?: number | null
  quantity?: string
  quantity_unit?: QuantityUnit
  captured_rate?: string | null
  override_rate?: string | null
  remarks?: string | null
}

export interface DailyConsumableLineIn {
  category: ConsumableCategory
  item_id?: number | null
  item_code?: string | null
  item_name?: string | null
  quantity?: string
  uom?: string | null
  currency?: string | null
  captured_rate?: string | null
  override_rate?: string | null
  manual_amount?: string | null
  section_id?: number | null
  phase_id?: number | null
  sub_activity_id?: number | null
  remarks?: string | null
}

export interface DailyTangibleLineIn {
  tangible_id: number | null
  quantity?: string
  uom?: string | null
  currency?: string | null
  captured_rate?: string | null
  override_rate?: string | null
  remarks?: string | null
}

export interface DailyCostSaveIn {
  services: DailyServiceLineIn[]
  consumables: DailyConsumableLineIn[]
  tangibles: DailyTangibleLineIn[]
  remarks?: string | null
}

export interface DailyCostPreviewIn extends DailyCostSaveIn {
  well_id: number
  afe_id?: number | null
}

export interface DailyServiceLine {
  id: number
  service_id: number
  service_code: string | null
  service_name: string | null
  provider_type: string | null
  afe_line_id: number | null
  charging_basis: ChargingBasis
  charge_category: string
  section_id: number | null
  phase_id: number | null
  sub_activity_id: number | null
  sub_activity_display: string | null
  quantity: string
  quantity_unit: string
  captured_rate: string
  override_rate: string | null
  amount: string
  remarks: string | null
}

export interface DailyConsumableLine {
  id: number
  category: ConsumableCategory
  item_id: number | null
  item_code: string
  item_name: string
  quantity: string
  uom: string | null
  currency: string | null
  captured_rate: string
  override_rate: string | null
  manual_amount: string | null
  amount: string
  section_id: number | null
  phase_id: number | null
  sub_activity_id: number | null
  sub_activity_display: string | null
  remarks: string | null
}

export interface DailyTangibleLine {
  id: number
  tangible_id: number
  tangible_code: string | null
  tangible_name: string | null
  quantity: string
  uom: string | null
  currency: string | null
  captured_rate: string
  override_rate: string | null
  amount: string
  remarks: string | null
}

export interface DailyCostEntry {
  id: number
  daily_cost_code: string
  rig_id: number
  well_id: number
  cost_date: string
  afe_id: number | null
  afe_code: string | null
  remarks: string | null
  status: DailyStatus
  submitted_at: string | null
  reconciliation_status: ReconciliationStatus
  reconciliation_ref: string | null
  reconciled_at: string | null
  is_deleted: boolean
  deleted_at: string | null
  created_at: string | null
  updated_at: string | null
  rig_code: string | null
  rig_name: string | null
  rig_display: string | null
  well_code: string | null
  well_name: string | null
  well_display: string | null
  service_count: number
  consumable_count: number
  tangible_count: number
  service_total: string
  consumable_total: string
  tangible_total: string
  total_cost: string
}

export interface SummaryRow {
  group: string
  amount: string
}

export interface DailyCostDay {
  entry: DailyCostEntry
  well_configuration: WellConfiguration | null
  services: DailyServiceLine[]
  consumables: DailyConsumableLine[]
  tangibles: DailyTangibleLine[]
  summary: SummaryRow[]
  grand_total: string
  warnings: string[]
}

/** One priced line of the server preview, matched back to the row on screen. */
export interface PreviewLine {
  line_id: number | null
  code: string
  name: string
  amount: string
  warnings: string[]
}

export interface DailyCostPreview {
  services: PreviewLine[]
  consumables: PreviewLine[]
  tangibles: PreviewLine[]
  summary: SummaryRow[]
  grand_total: string
  warnings: string[]
}

// ---------------------------------------------------------------------------
// Catalogue lookups the panels pick from
// ---------------------------------------------------------------------------

export interface MudChemicalOption {
  id: number
  chemical_code: string
  chemical_name: string
  current_rate: string | null
  uom: string | null
  currency: string | null
  [key: string]: unknown
}

export interface DrillBitOption {
  id: number
  bit_code: string
  bit_name: string
  final_cost: string | null
  uom: string | null
  currency: string | null
  [key: string]: unknown
}

export interface TangibleOption {
  id: number
  tangible_code: string
  tangible_name: string
  final_cost: string | null
  uom: string | null
  currency: string | null
  [key: string]: unknown
}

export interface ServiceOption {
  id: number
  service_code: string
  service_name: string
  provider_type: string
  [key: string]: unknown
}

// ---------------------------------------------------------------------------
// Cost Analytics
// ---------------------------------------------------------------------------

export interface GroupComparison {
  group: string
  estimated: string
  actual: string
  balance: string
  variance_pct: string | null
  [key: string]: unknown
}

export interface CostForecast {
  actual_to_date: string
  estimated_total: string
  planned_days: string
  elapsed_days: string
  remaining_days: string
  burn_rate_per_day: string
  forecast_at_completion: string
  variance: string
  variance_pct: string | null
  balance_at_completion: string
  basis: string
}

export interface DepthCostPoint {
  depth: string
  section_id: number
  section_label: string
  estimated_cumulative: string
  actual_cumulative: string
  estimated_section: string
  actual_section: string
  variance: string
}

export interface WellAnalyticsSummary {
  well_id: number
  well_code: string
  well_name: string
  rig_id: number | null
  rig_code: string | null
  rig_name: string | null
  well_status: string
  depth_unit: string
  afe_count: number
  estimated_total: string
  estimated_services: string
  estimated_consumables: string
  estimated_tangibles: string
  actual_total: string
  actual_services: string
  actual_consumables: string
  actual_tangibles: string
  balance: string
  utilisation: string | null
  reconciled_total: string
  unreconciled_total: string
  planned_days: string
  elapsed_days: string
  days_with_cost: number
  first_cost_date: string | null
  last_cost_date: string | null
  forecast_at_completion: string
  forecast_variance: string
}

export interface DailyTrendPoint {
  cost_date: string
  amount: string
  cumulative: string
}

export interface DimensionRow {
  key: string
  label: string
  services: string
  consumables: string
  tangibles: string
  total: string
  estimated: string
  balance: string
}

export interface WellAnalytics {
  well: WellAnalyticsSummary
  afe_id: number | null
  afes: DailyCostAfeOption[]
  comparisons: GroupComparison[]
  forecast: CostForecast
  depth_series: DepthCostPoint[]
  depth_notes: string[]
  unattributed_actual: string
  dimensions: Record<string, DimensionRow[]>
  daily_trend: DailyTrendPoint[]
  warnings: string[]
}

// ---------------------------------------------------------------------------
// Cost Reports
// ---------------------------------------------------------------------------

export type ReportDimension =
  | 'date'
  | 'section'
  | 'phase'
  | 'activity'
  | 'sub_activity'
  | 'service'
  | 'charge_category'
  | 'consumable_category'
  | 'tangible'
  | 'well'

export interface ReportDimensionOption {
  dimension: ReportDimension
  title: string
}

export interface ReportRow {
  key: string
  label: string
  services: string
  consumables: string
  tangibles: string
  total: string
  estimated: string
  balance: string
  extra: Record<string, unknown>
}

export interface ReportTotals {
  services: string
  consumables: string
  tangibles: string
  total: string
  estimated: string
  balance: string
}

export interface CostReport {
  dimension: ReportDimension
  title: string
  filters: Record<string, unknown>
  rows: ReportRow[]
  totals: ReportTotals
  generated_at: string | null
}

/** One cost line behind a report row — the drill-through itself. */
export interface ReportLine {
  cost_date: string
  daily_cost_code: string
  well: string
  cost_group: string
  category: string
  code: string
  name: string
  section: string
  phase: string
  activity: string
  sub_activity: string
  quantity: string
  unit: string
  rate: string
  amount: string
  remarks: string
  status: DailyStatus
}

export interface ReportLineBundle {
  dimension: ReportDimension
  key: string | null
  line_count: number
  total: number
  lines: ReportLine[]
}

File: backend/app/schemas/daily_cost.py
"""Pydantic schemas for Daily Costs (and the Cost Analytics / Reports pages)."""

from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.master_data import BlankStr, FlagBool
from app.schemas.rig_well import WellConfigurationOut

DailyCostStatus = Literal["draft", "submitted"]
ChargingBasis = Literal["Daily Rate", "Per Service Rate", "Per Section Rate"]
QuantityUnit = Literal["days", "hours"]
ConsumableCategory = Literal["mud_chemical", "fuel", "cement_additive", "drill_bit"]
CostGroup = Literal["Services", "Consumables", "Tangibles"]
ReportDimension = Literal[
    "date",
    "section",
    "phase",
    "activity",
    "sub_activity",
    "service",
    "charge_category",
    "consumable_category",
    "tangible",
    "well",
]


# ---------------------------------------------------------------------------
# Inputs
# ---------------------------------------------------------------------------


class DailyCostEntryIn(BaseModel):
    """Create the day sheet for one rig + well + date."""

    well_id: int = Field(..., description="Well the day belongs to (the rig follows the well)")
    cost_date: date
    afe_id: int | None = Field(None, description="AFE that supplies the rate card")
    remarks: str | None = None


class DailyCostEntryUpdate(BaseModel):
    cost_date: date | None = None
    afe_id: int | None = None
    remarks: str | None = None


class DailyServiceLineIn(BaseModel):
    """One service line of the day.

    ``charging_basis`` and ``captured_rate`` are optional on purpose: when the
    service exists on the day's AFE the server takes both from that AFE line
    (the AFE owns the rate card), so the browser never decides money. They are
    only honoured when the service is not on the AFE and the user typed the
    rate by hand.
    """

    service_id: int
    charging_basis: ChargingBasis | None = None
    charge_category: str | None = None
    afe_line_id: int | None = None
    section_id: int | None = None
    phase_id: int | None = None
    sub_activity_id: int | None = None
    quantity: Decimal = Field(Decimal("0"), ge=0)
    quantity_unit: QuantityUnit = "hours"
    captured_rate: Decimal | None = None
    override_rate: Decimal | None = Field(None, ge=0)
    remarks: str | None = None


class DailyConsumableLineIn(BaseModel):
    category: ConsumableCategory = "mud_chemical"
    item_id: int | None = None
    item_code: str | None = None
    item_name: str | None = None
    quantity: Decimal = Field(Decimal("0"), ge=0)
    uom: str | None = None
    currency: str | None = None
    captured_rate: Decimal | None = None
    override_rate: Decimal | None = Field(None, ge=0)
    #: Cement additives: the total consumption cost for the chosen scope.
    manual_amount: Decimal | None = Field(None, ge=0)
    section_id: int | None = None
    phase_id: int | None = None
    sub_activity_id: int | None = None
    remarks: str | None = None


class DailyTangibleLineIn(BaseModel):
    tangible_id: int
    quantity: Decimal = Field(Decimal("1"), ge=0)
    uom: str | None = None
    currency: str | None = None
    captured_rate: Decimal | None = None
    override_rate: Decimal | None = Field(None, ge=0)
    remarks: str | None = None


class DailyCostSaveIn(BaseModel):
    """The whole day, saved atomically."""

    services: list[DailyServiceLineIn] = Field(default_factory=list)
    consumables: list[DailyConsumableLineIn] = Field(default_factory=list)
    tangibles: list[DailyTangibleLineIn] = Field(default_factory=list)
    remarks: str | None = None


class DailyCostPreviewIn(DailyCostSaveIn):
    """Preview payload — priced before the day exists."""

    well_id: int
    afe_id: int | None = None


class DailyStatusIn(BaseModel):
    action: Literal["submit", "reopen"]
    remarks: str | None = None


# ---------------------------------------------------------------------------
# Rate card (what the AFE says about each service)
# ---------------------------------------------------------------------------


class RateCardServiceOut(BaseModel):
    """One service of the AFE rate card, as the daily page needs it."""

    service_id: int
    afe_line_id: int | None = None
    service_code: str = ""
    service_name: str = ""
    provider_type: str = ""
    charging_basis: str = "Daily Rate"
    per_service_amount: Decimal = Decimal("0")
    section_id: int | None = None
    phase_id: int | None = None
    rates: list[dict[str, object]] = Field(default_factory=list)
    section_rates: list[dict[str, object]] = Field(default_factory=list)


class DailyCostContextOut(BaseModel):
    """Everything the daily page needs once the rig/well/date are picked."""

    well_id: int
    well_code: str = ""
    well_name: str = ""
    rig_id: int | None = None
    rig_code: str | None = None
    rig_name: str | None = None
    depth_unit: str = "m"
    well_configuration: WellConfigurationOut | None = None
    afes: list[dict[str, object]] = Field(default_factory=list)
    sub_activities: list[dict[str, object]] = Field(default_factory=list)
    rate_card: list[RateCardServiceOut] = Field(default_factory=list)
    afe_id: int | None = None
    #: Fuel unit rate captured on the selected AFE — the only consumable the
    #: daily page reads its rate from the AFE for.
    fuel_rate: Decimal = Decimal("0")
    #: The selected AFE's estimated total, so the sheet can show what is left.
    afe_estimated_total: Decimal = Decimal("0")
    warnings: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Read models
# ---------------------------------------------------------------------------


class DailyLineWarningOut(BaseModel):
    line_id: int | None = None
    message: str


class DailyServiceLineOut(BaseModel):
    id: int
    service_id: int
    service_code: str | None = None
    service_name: str | None = None
    provider_type: str | None = None
    afe_line_id: int | None = None
    charging_basis: str
    charge_category: str
    section_id: int | None = None
    phase_id: int | None = None
    sub_activity_id: int | None = None
    sub_activity_display: str | None = None
    quantity: Decimal = Decimal("0")
    quantity_unit: str = "hours"
    captured_rate: Decimal = Decimal("0")
    override_rate: Decimal | None = None
    amount: Decimal = Decimal("0")
    remarks: str | None = None

    model_config = ConfigDict(from_attributes=True)


class DailyConsumableLineOut(BaseModel):
    id: int
    category: str
    item_id: int | None = None
    item_code: str
    item_name: str
    quantity: Decimal = Decimal("0")
    uom: str | None = None
    currency: str | None = None
    captured_rate: Decimal = Decimal("0")
    override_rate: Decimal | None = None
    manual_amount: Decimal | None = None
    amount: Decimal = Decimal("0")
    section_id: int | None = None
    phase_id: int | None = None
    sub_activity_id: int | None = None
    sub_activity_display: str | None = None
    remarks: str | None = None

    model_config = ConfigDict(from_attributes=True)


class DailyTangibleLineOut(BaseModel):
    id: int
    tangible_id: int
    tangible_code: str | None = None
    tangible_name: str | None = None
    quantity: Decimal = Decimal("1")
    uom: str | None = None
    currency: str | None = None
    captured_rate: Decimal = Decimal("0")
    override_rate: Decimal | None = None
    amount: Decimal = Decimal("0")
    remarks: str | None = None

    model_config = ConfigDict(from_attributes=True)


class DailyCostEntryOut(BaseModel):
    """The day's header with its three group totals — the list read model."""

    id: int
    daily_cost_code: BlankStr = ""
    rig_id: int
    well_id: int
    cost_date: date
    afe_id: int | None = None
    afe_code: str | None = None
    remarks: str | None = None
    status: BlankStr = "draft"
    submitted_at: datetime | None = None
    reconciliation_status: BlankStr = "pending"
    reconciliation_ref: str | None = None
    reconciled_at: datetime | None = None
    is_deleted: FlagBool = False
    deleted_at: datetime | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    rig_code: str | None = None
    rig_name: str | None = None
    rig_display: str | None = None
    well_code: str | None = None
    well_name: str | None = None
    well_display: str | None = None
    service_count: int = 0
    consumable_count: int = 0
    tangible_count: int = 0
    service_total: Decimal = Decimal("0")
    consumable_total: Decimal = Decimal("0")
    tangible_total: Decimal = Decimal("0")
    total_cost: Decimal = Decimal("0")

    model_config = ConfigDict(from_attributes=True)


class DailyCostDayOut(BaseModel):
    """The full day: header, configuration, lines, totals and warnings."""

    entry: DailyCostEntryOut
    well_configuration: WellConfigurationOut | None = None
    services: list[DailyServiceLineOut] = Field(default_factory=list)
    consumables: list[DailyConsumableLineOut] = Field(default_factory=list)
    tangibles: list[DailyTangibleLineOut] = Field(default_factory=list)
    summary: list[dict[str, object]] = Field(default_factory=list)
    grand_total: Decimal = Decimal("0")
    warnings: list[str] = Field(default_factory=list)


class DailyCostPreviewOut(BaseModel):
    """Live totals for the rows on screen — the same engine that saves them."""

    services: list[dict[str, object]] = Field(default_factory=list)
    consumables: list[dict[str, object]] = Field(default_factory=list)
    tangibles: list[dict[str, object]] = Field(default_factory=list)
    summary: list[dict[str, object]] = Field(default_factory=list)
    grand_total: Decimal = Decimal("0")
    warnings: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Analytics / reports
# ---------------------------------------------------------------------------


class GroupComparisonOut(BaseModel):
    group: str
    estimated: Decimal = Decimal("0")
    actual: Decimal = Decimal("0")
    reconciled: Decimal = Decimal("0")
    unreconciled: Decimal = Decimal("0")
    balance: Decimal = Decimal("0")
    utilisation: Decimal | None = None


class CostForecastOut(BaseModel):
    actual_to_date: Decimal = Decimal("0")
    estimated_total: Decimal = Decimal("0")
    planned_days: Decimal = Decimal("0")
    elapsed_days: Decimal = Decimal("0")
    remaining_days: Decimal = Decimal("0")
    burn_rate_per_day: Decimal = Decimal("0")
    forecast_at_completion: Decimal = Decimal("0")
    variance: Decimal = Decimal("0")
    variance_pct: Decimal | None = None
    balance_at_completion: Decimal = Decimal("0")
    basis: str = ""


class DepthCostPointOut(BaseModel):
    depth: Decimal
    section_id: int | None = None
    section_label: str = ""
    estimated_cumulative: Decimal = Decimal("0")
    actual_cumulative: Decimal = Decimal("0")
    estimated_section: Decimal = Decimal("0")
    actual_section: Decimal = Decimal("0")
    variance: Decimal = Decimal("0")


class WellAnalyticsSummaryOut(BaseModel):
    """One row of the Cost Analytics table."""

    well_id: int
    well_code: str = ""
    well_name: str = ""
    rig_id: int | None = None
    rig_code: str | None = None
    rig_name: str | None = None
    well_status: str = ""
    depth_unit: str = "m"
    afe_count: int = 0
    estimated_total: Decimal = Decimal("0")
    estimated_services: Decimal = Decimal("0")
    estimated_consumables: Decimal = Decimal("0")
    estimated_tangibles: Decimal = Decimal("0")
    actual_total: Decimal = Decimal("0")
    actual_services: Decimal = Decimal("0")
    actual_consumables: Decimal = Decimal("0")
    actual_tangibles: Decimal = Decimal("0")
    balance: Decimal = Decimal("0")
    utilisation: Decimal | None = None
    reconciled_total: Decimal = Decimal("0")
    unreconciled_total: Decimal = Decimal("0")
    planned_days: Decimal = Decimal("0")
    elapsed_days: Decimal = Decimal("0")
    days_with_cost: int = 0
    first_cost_date: date | None = None
    last_cost_date: date | None = None
    forecast_at_completion: Decimal = Decimal("0")
    forecast_variance: Decimal = Decimal("0")


class WellAnalyticsOut(BaseModel):
    """The Cost Analytics detail for one well."""

    well: WellAnalyticsSummaryOut
    afe_id: int | None = None
    afes: list[dict[str, object]] = Field(default_factory=list)
    comparisons: list[GroupComparisonOut] = Field(default_factory=list)
    forecast: CostForecastOut
    depth_series: list[DepthCostPointOut] = Field(default_factory=list)
    depth_notes: list[str] = Field(default_factory=list)
    unattributed_actual: Decimal = Decimal("0")
    dimensions: dict[str, list[dict[str, object]]] = Field(default_factory=dict)
    daily_trend: list[dict[str, object]] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class ReportRowOut(BaseModel):
    """One drill-through row of the Reports page."""

    key: str
    label: str
    services: Decimal = Decimal("0")
    consumables: Decimal = Decimal("0")
    tangibles: Decimal = Decimal("0")
    total: Decimal = Decimal("0")
    estimated: Decimal = Decimal("0")
    balance: Decimal = Decimal("0")
    extra: dict[str, object] = Field(default_factory=dict)


class ReportOut(BaseModel):
    dimension: str
    title: str
    filters: dict[str, object] = Field(default_factory=dict)
    rows: list[ReportRowOut] = Field(default_factory=list)
    totals: dict[str, Decimal] = Field(default_factory=dict)
    generated_at: datetime | None = None
